src/map.c

```diff
@@ -3,752 +3,641 @@
+    } while (0)
+
+#define rb_node_rotate_right(x_type, x_field, x_node, r_node)    \
+    do {                                                         \
+        (r_node) = rb_node_get_left((x_node));                   \
+        rb_node_set_left((x_node), rb_node_get_right((r_node))); \
+        rb_node_set_right((r_node), (x_node));                   \
+    } while (0)
+
+typedef struct {
+    map_node_t *node;
+    map_cmp_t cmp;
+} rb_path_entry_t;
+
+static inline map_node_t *rb_search(map_t rb, const map_node_t *node)
 {
-    /* Rotate to the left according to parent */
-    parent = map_rotate_left(obj, parent);
-
-    /* Refigure out the identity */
-    node = parent->left;
-    grandparent = rb_parent(parent);
-    uncle =
-        (grandparent->left == parent) ? grandparent->right : grandparent->left;
-
-    /* Apply left-left case */
-    map_l_l(obj, node, parent, grandparent, uncle);
+    map_node_t *ret = rb->root;
+    while (ret) {
+        map_cmp_t cmp = (rb->cmp)(node->key, ret->key);
+        switch (cmp) {
+        case _CMP_EQUAL:
+            return ret;
+        case _CMP_LESS:
+            ret = rb_node_get_left(ret);
+            break;
+        case _CMP_GREATER:
+            ret = rb_node_get_right(ret);
+            break;
+        default:
+            __UNREACHABLE;
+            break;
+        }
+    }
+    return ret;
 }
 
-static void map_r_r(map_t obj,
-                    map_node_t *node UNUSED,
-                    map_node_t *parent UNUSED,
-                    map_node_t *grandparent,
-                    map_node_t *uncle UNUSED)
+static void rb_insert(map_t rb, map_node_t *node)
 {
-    /* Rotate to the left according to grandparent */
-    grandparent = map_rotate_left(obj, grandparent);
-
-    /* Swap grandparent and uncle's colors */
-    map_color_t c1 = rb_color(grandparent), c2 = rb_color(grandparent->left);
-
-    rb_set_color(grandparent, c2);
-    rb_set_color(grandparent->left, c1);
-}
+    rb_path_entry_t path[RB_MAX_DEPTH];
+    rb_path_entry_t *pathp;
+    rb_node_init(node);
 
-static void map_r_l(map_t obj,
-                    map_node_t *node,
-                    map_node_t *parent,
-                    map_node_t *grandparent,
-                    map_node_t *uncle)
-{
-    /* Rotate to the right according to parent */
-    parent = map_rotate_right(obj, parent);
+    /* Wind.
+     * Traverse through red-black tree node and find the search target node.
+     * The path from root to seach target is recorded in pathp[i].
+     */
+    path->node = rb->root;
+    for (pathp = path; pathp->node; pathp++) {
+        map_cmp_t cmp = pathp->cmp = (rb->cmp)(node->key, pathp->node->key);
+        switch (cmp) {
+        case _CMP_LESS:
+            pathp[1].node = rb_node_get_left(pathp->node);
+            break;
+        case _CMP_GREATER:
+            pathp[1].node = rb_node_get_right(pathp->node);
+            break;
+        default:
+            /* igore duplicate key */
```

Comment:
Shall we add a proper `__UNREACHABLE` here?

Reply:
Yes, after consideration, we should add this. And I have added.